### PR TITLE
feat(artifacts): model registry - adds the ability to register artifacts from the sdk

### PR DESCRIPTION
**Pull request by ibindlish**
Fixes [WB-10443](https://wandb.atlassian.net/browse/WB-10443)
Backend dependency: https://github.com/wandb/core/pull/10791

Description
-----------
Model registry
Allows users to created registered artifacts of type "model" from the sdk via `run.register_artifact()`
Only supported for cli versions >= `update based on backend changes`

Testing
-------
Tested against the wandb backend using the local-testcontainer framework. 
`test_public_api.py :: test_create_registered_model`

### CODE REVIEW

**Review by tssweeney — CHANGES_REQUESTED** [review]
This is really great work - nice job! There are a few notes:
1. Can you work with @vwrj to come up with a more generic method name? The term `registered artifact` is a bit model-specific and I bet we will iterate on it some more. Something closer to the true API call, like `create_collection`.
2. Since the method is not intended to be ran more than once, it should be attached to the API object instead of the run. A good rule of thumb is a run is like a function - its config is the parameters and its artifacts + history logs are the output. Any implementation should be able to run over and over without side effects. 
3. Can we schedule a quick followup PR to make it so when a user calls `link_artifact`, if the collection does not exist, one is created? This would be a pure core PR, modifying the `linkArtifact` resolver. I think this is the more common case that would effect most users.

**Comment by ibindlish** [comment]
> 3\. make it so when a user calls `link_artifact`, if the collection does not exist, one is created

Regarding the function name, should we set it as `portfolio` instead of `collection` to differentiate between ordered(or versioned) artifact collections and grouped artifacts? Or do we not intend to make this distinction customer facing?

**Comment by ibindlish** [comment]
Any suggestions for the function name, unless `create_artifact_portfolio` is ok?

File: wandb/apis/public.py [hunk 1]
```diff
@@ -5192,3 +5226,157 @@ def call(
             cuda=cuda,
         )
         return queued_run
+
+
+class ArtifactPortfolio(Attrs):
```
Comment (tssweeney) [inline wandb/apis/public.py:5231]: How might we restructure this to share common features between Portfolios and Sequences? Notice in `class ArtifactCollection` (which uses Sequence GQL endpoints), we have a methods like `load`, `version`, etc... As a user, they should not need to think about if something is a sequence or a portfolio. Regardless, for example, they should be able to get the versions. 

This relates to the bigger picture we have been talking about - anytime you find yourself writing something Portfolio or Sequence specific, it is probably too specific. the only differences between the two are:
1. A single field in the DB
2. An Artifact version always belongs to 1 and only 1 sequence, but may belong to 0-to-many portfolios. 

The second rule is just a business rule on memberships. Every other class/method/api we build should be the same between the two.

Comment (ibindlish) [reply wandb/apis/public.py:5231]: Moving this to be handled by the LinkArtifact resolver on the backend instead. Thanks for looking into this Tim!

**Pull request closed without merging**